src/lib/tuning/greedy_selector.cpp

```diff
@@ -0,0 +1,157 @@
+ * Otherwise the end() iterator is returned, that the constraints are not satisfied
+ * by any continuous subsequence starting at sorted_choices.begin().
+ */
+const std::list<std::shared_ptr<TuningChoice>>::const_iterator sacrifice_choices(
+    const std::list<std::shared_ptr<TuningChoice>>& sorted_choices, float required_cost_delta,
+    float acceptible_desirability_delta = -std::numeric_limits<float>::infinity()) {
+  auto desirability_delta = 0.0f;
+  auto cost_delta = 0.0f;
+
+  auto choice = sorted_choices.cbegin();
+  while (choice != sorted_choices.cend()) {
+    if (cost_delta <= required_cost_delta) {
+      break;
+    }
+    desirability_delta += (*choice)->reject_desirability();
+    cost_delta += (*choice)->reject_cost();
+    ++choice;
+  }
+
+  if (cost_delta <= required_cost_delta && desirability_delta >= acceptible_desirability_delta) {
+    return choice;
+  }
+  return sorted_choices.cend();
+}
+
+std::vector<std::shared_ptr<TuningOperation>> GreedySelector::select(
+    const std::vector<std::shared_ptr<TuningChoice>>& choices, float cost_budget) {
+  std::vector<std::shared_ptr<TuningOperation>> operations;
+  operations.reserve(choices.size());
+  // Assumption: cost() >= 0 ==> accept_cost() >= 0 && current_cost() >= 0 && reject_cost() <= 0
+
+  // Accumulate absolute cost balance from currently chosen choices.
+  float cost_balance = 0.0f;
+  for (const auto& choice : choices) {
+    cost_balance += choice->current_cost();
+  }
+
+  // Desirability balance is always relative to current system state.
+  float desirability_balance = 0.0f;
+
+  // Sort choices:
+  //  1) by reject_cost() ascending
+  //  2) by accept_desirability() ascending stable
+  // The most desirable choice is at back() and among equaly desirable choices
+  // the least costly choices to accept are nearer back().
+  std::list<std::shared_ptr<TuningChoice>> sorted_choices(choices.cbegin(), choices.cend());
+
+  sorted_choices.sort([](std::shared_ptr<TuningChoice> lhs, std::shared_ptr<TuningChoice> rhs) {
+    return lhs->reject_cost() < rhs->reject_cost();
+  });
+  sorted_choices.sort([](std::shared_ptr<TuningChoice> lhs, std::shared_ptr<TuningChoice> rhs) {
+    return lhs->accept_desirability() < rhs->accept_desirability();
+  });
+
+  // If current state exceeds cost_budget,
+  // reject the least desirable choices to reduce cost_balance.
+  if (cost_balance > cost_budget) {
+    LOG_INFO("Cost balance of " << cost_balance << " exceeds budget of " << cost_budget);
+    const auto sacrifice_until = sacrifice_choices(sorted_choices, cost_budget - cost_balance);
+    if (sacrifice_until == sorted_choices.cend()) {
+      LOG_WARN("Cost budget is impossible to maintain. No Operations are performed.");
+      return operations;
+    }
+    for (auto choice = sorted_choices.cbegin(); choice != sacrifice_until; ++choice) {
+      LOG_DEBUG(" ! Reject " << **choice << " to reduce cost balance");
+      operations.push_back((*choice)->reject());
```

Comment:
Starting to understand a bit more. I think your use of „reject“ is a bit misleading. For me, rejecting an option means that I decide not to act. What you are doing goes more into „undoing“ a previously made decision. Or maybe „perform/revert/ignore“ instead of „accept/reject/sacrifice“.

Is a `reject_cost` the cost that is incurred by reverting a decision (e.g., the cost of deallocating an index) or is it the runtime cost incurred because we don‘t have the index anymore?

I think the naming should be revisited. Starting with „choice“ (which for me is an option that was already decided on), then accept/reject, and finally cost/desirability.

Reply:
For us, the term choice was the best fit for something that might or might not already be incorporated in the current system state. Depending on its presence, an accept would become a keep or do/perform and a reject an undo/revert or an ignore operation. The fact that both accepting and rejecting it might (alternatively) lead to a nontrivial TuningOperation seems not to be represented by the alterative terms we discussed.

Also to my mind the term choice can refer equally well to a choice that is already made or still open.

Reply:
Let's do the "choice" discussion in the other comment.